### PR TITLE
Merge fad/in-app-feedback into fad/next to prepare for release

### PR DESCRIPTION
**Pull request by lfkellogg**

### CODE REVIEW

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FeedbackActivity.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AppCompatActivity;
+import com.google.firebase.annotations.concurrent.Blocking;
+import com.google.firebase.annotations.concurrent.UiThread;
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+
+/** Activity for tester to compose and submit feedback. */
+public class FeedbackActivity extends AppCompatActivity {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FeedbackActivity.java:45]: Please review

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java [hunk 2]
```diff
@@ -25,12 +29,16 @@
...
 import com.google.firebase.appdistribution.FirebaseAppDistributionException;
 import java.util.ArrayDeque;
 import java.util.Queue;
 import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Singleton;
 
-class FirebaseAppDistributionLifecycleNotifier implements Application.ActivityLifecycleCallbacks {
+@Singleton // Only one lifecycle notifier is required across the entire app
+class FirebaseAppDistributionLifecycleNotifier {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java:41]: Please review. This was updated to support an "ignored" class when getting the current activity. This is so we can take a screenshot of the previous activity while `TakeScreenshotAndStartFeedbackActivity` is active.

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionNotificationsManager.java [hunk 3]
```diff
@@ -14,96 +14,262 @@
...
+import com.google.firebase.appdistribution.impl.FirebaseAppDistributionLifecycleNotifier.OnActivityPausedListener;
+import com.google.firebase.appdistribution.impl.FirebaseAppDistributionLifecycleNotifier.OnActivityResumedListener;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+class FirebaseAppDistributionNotificationsManager
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionNotificationsManager.java:49]: Please review

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ImageUtils.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.graphics.Rect;
+import android.net.Uri;
+import androidx.annotation.Nullable;
+import com.google.auto.value.AutoValue;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class ImageUtils {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ImageUtils.java:27]: Please review

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ReleaseIdentifier.java [hunk 5]
```diff
@@ -193,4 +213,25 @@ private static byte[] arrayListToByteArray(ArrayList<Byte> list) {
     }
     return result;
   }
+
+  private static boolean developmentModeEnabled() {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ReleaseIdentifier.java:217]: Please review

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ScreenshotTaker.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+import java.io.BufferedOutputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/** A class that takes screenshots of the host app. */
+@Singleton // All instances store images at the same URI, so there should really only be one
+class ScreenshotTaker {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ScreenshotTaker.java:44]: Please review

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/TakeScreenshotAndStartFeedbackActivity.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+
+package com.google.firebase.appdistribution.impl;
+
+import android.app.Activity;
+import android.os.Bundle;
+import com.google.firebase.appdistribution.FirebaseAppDistribution;
+import javax.inject.Inject;
+
+public class TakeScreenshotAndStartFeedbackActivity extends Activity {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/TakeScreenshotAndStartFeedbackActivity.java:22]: Please review

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/TaskCompletionSourceCache.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import com.google.firebase.concurrent.FirebaseExecutors;
+import java.util.concurrent.Executor;
+
+/**
+ * A cache for a {@link TaskCompletionSource}, for use in cases where we only ever want one active
+ * task at a time for a particular operation.
+ *
+ * <p>This is equivalent to {@link UpdateTaskCache} but for Tasks.
+ */
+class TaskCompletionSourceCache<T> {
```
Comment (lfkellogg) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/TaskCompletionSourceCache.java:29]: Please review

**Review by lfkellogg — COMMENTED** [review]
@vkryachko @rlazo Please take a look at the sections I've marked with "Please review" which cover the most important and/or novel aspects of the in-app feedback feature.

**Review by tagboola — APPROVED** [review]

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java [hunk 9]
```diff
@@ -123,6 +153,56 @@ <T> Task<T> applyToForegroundActivity(ActivityFunction<T> function) {
...
+   * the activity has been destroyed.
+   *
+   * <p>The function will always be called on the UI thread.
+   */
+  <T, A extends Activity> Task<T> applyToNullableForegroundActivity(
+      Class<A> classToIgnore, NullableActivityFunction<T> function) {
+    return getForegroundActivity(classToIgnore)
+        .onSuccessTask(
+            // Use direct executor to ensure the consumer is called while Activity is in foreground
+            DIRECT_EXECUTOR,
```
Comment (rlazo) [inline firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java:171]: FYI there's a `FirebaseExecutors.directExecutor()`

**Review by rlazo — APPROVED** [review]

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java (hunk 9 above)
Comment (lfkellogg) [reply firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java]: Thanks! Replaced

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java [hunk 10]
```diff
@@ -101,16 +118,29 @@ interface OnActivityDestroyedListener {
...
    * will complete immediately after the function is applied.
    *
-   * <p>The function will be called immediately once the activity is available. This may be main
-   * thread or the calling thread, depending on whether or not there is already a foreground
-   * activity available when this method is called.
+   * <p>The function will always be called on the UI thread.
    */
   <T> Task<T> applyToForegroundActivity(ActivityFunction<T> function) {
-    return getForegroundActivity()
+    return getForegroundActivity(null)
```
[the remark this answers is not in the document]
Comment (lfkellogg) [reply firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FirebaseAppDistributionLifecycleNotifier.java]: Actually, there's no reason to not use `getForegroundActivity()`. Went back to using that, and also added a parameter comment on that method's usage: `getForegroundActivity(/* classToIgnore= */ null)`

File: firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FeedbackActivity.java [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+          FeedbackTrigger.fromString(savedInstanceState.getString(FEEDBACK_TRIGGER_KEY));
+      String screenshotUriString = savedInstanceState.getString(SCREENSHOT_URI_KEY);
+      if (screenshotUriString != null) {
+        screenshotUri = Uri.parse(screenshotUriString);
+      }
+    } else {
+      releaseName = getIntent().getStringExtra(RELEASE_NAME_KEY);
+      additionalFormText = getIntent().getCharSequenceExtra(ADDITIONAL_FORM_TEXT_KEY);
+      feedbackTrigger =
+          FeedbackTrigger.fromString(getIntent().getStringExtra(FEEDBACK_TRIGGER_KEY));
```
[the remark this answers is not in the document]
Comment (lfkellogg) [reply firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/FeedbackActivity.java:91]: Good call. Follow up: https://github.com/firebase/firebase-android-sdk/pull/4689

**Pull request merged**